src/main/java/org/sonarlint/intellij/core/ServerIssueUpdater.java

```diff
@@ -0,0 +1,146 @@
+import org.sonarlint.intellij.util.SonarLintUtils;
+import org.sonarsource.sonarlint.core.client.api.connected.ConnectedSonarLintEngine;
+import org.sonarsource.sonarlint.core.client.api.connected.ServerIssue;
+import org.sonarsource.sonarlint.core.client.api.exceptions.SonarLintWrappedException;
+
+public class ServerIssueUpdater extends AbstractProjectComponent {
+
+  private static final Logger LOGGER = Logger.getInstance(ServerIssueUpdater.class);
+
+  private static final int THREADS_NUM = 5;
+  private static final int QUEUE_LIMIT = 20;
+
+  private ExecutorService executorService;
+
+  private final IssueStore store;
+
+  public ServerIssueUpdater(Project project, IssueStore store) {
+    super(project);
+    this.store = store;
+  }
+
+  public void trackServerIssues(VirtualFile virtualFile) {
+    SonarLintProjectSettings projectSettings = SonarLintUtils.get(myProject, SonarLintProjectSettings.class);
+    if (!projectSettings.isBindingEnabled()) {
+      // not in connected mode
+      return;
+    }
+
+    String serverId = projectSettings.getServerId();
+    String moduleKey = projectSettings.getProjectKey();
+    if (serverId == null || moduleKey == null) {
+      // not bound to SQ project
+      return;
+    }
+
+    ConnectedSonarLintEngine engine = SonarLintUtils.get(SonarLintServerManager.class).getConnectedEngine(serverId);
+
+    String relativePath = getRelativePath(virtualFile);
+
+    fetchAndMatchServerIssues(virtualFile, engine, moduleKey, relativePath);
+  }
+
+  private void fetchAndMatchServerIssues(VirtualFile virtualFile, ConnectedSonarLintEngine engine, String moduleKey, String relativePath) {
+    Runnable task = () -> {
+      Iterator<ServerIssue> serverIssues = fetchServerIssues(engine, moduleKey, relativePath);
+
+      Collection<IssuePointer> serverIssuePointers = toStream(serverIssues).map(ServerIssuePointer::new).collect(Collectors.toList());
+
+      if (!serverIssuePointers.isEmpty()) {
+        store.matchWithServerIssues(virtualFile, serverIssuePointers);
+      }
+    };
+    try {
+      this.executorService.submit(task);
+    } catch (RejectedExecutionException e) {
+      LOGGER.debug("fetch and match server issues rejected for moduleKey=" + moduleKey + ", filepath=" + relativePath, e);
+    }
+  }
+
+  private static <T> Stream<T> toStream(Iterator<T> iterator) {
+    Iterable<T> iterable = () -> iterator;
+    return StreamSupport.stream(iterable.spliterator(), false);
+  }
+
+  private Iterator<ServerIssue> fetchServerIssues(ConnectedSonarLintEngine engine, String moduleKey, String relativePath) {
+    try {
+      LOGGER.debug("fetchServerIssues moduleKey=" + moduleKey + ", filepath=" + relativePath);
+      return engine.downloadServerIssues(moduleKey, relativePath);
+    } catch (SonarLintWrappedException e) {
+      SonarLintConsole.get(myProject).error("could not download server issues", e);
+      return engine.getServerIssues(moduleKey, relativePath);
+    } catch (Throwable t) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Catch Exception instead of Throwable. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1181)